File: src/ormai/policy/budgets.py
"""
Budget enforcement and complexity scoring.

Budgets prevent runaway queries by limiting rows, includes, fields,
and query complexity.
"""

from ormai.core.dsl import FilterClause, IncludeClause, QueryRequest
from ormai.core.errors import QueryBudgetExceededError
from ormai.policy.models import Budget


class ComplexityScorer:
    """
    Scores query complexity based on various factors.

    Higher scores indicate more complex (and potentially expensive) queries.
    """

    # Default weights for complexity factors
    DEFAULT_WEIGHTS = {
        "base": 1,
        "per_field": 1,
        "per_filter": 2,
        "per_include": 10,
        "per_order": 1,
        "string_filter": 3,  # contains, startswith, endswith
        "in_filter": 2,  # per item in IN clause
        "between_filter": 2,
    }

    def __init__(self, weights: dict[str, int] | None = None) -> None:
        self.weights = {**self.DEFAULT_WEIGHTS, **(weights or {})}

    def score(self, request: QueryRequest) -> int:
        """
        Calculate complexity score for a query request.

        Returns an integer score where higher = more complex.
        """
        score = self.weights["base"]

        # Field selection
        if request.select:
            score += len(request.select) * self.weights["per_field"]

        # Filters
        if request.where:
            for f in request.where:
                score += self._score_filter(f)

        # Includes
        if request.include:
            for inc in request.include:
                score += self._score_include(inc)

        # Ordering
        if request.order_by:
            score += len(request.order_by) * self.weights["per_order"]

        return score

    def _score_filter(self, filter_clause: FilterClause) -> int:
        """Score a single filter clause."""
        score = self.weights["per_filter"]

        # String operations are more expensive
        if filter_clause.op in ("contains", "startswith", "endswith"):
            score += self.weights["string_filter"]

        # IN clauses scale with list size
        if filter_clause.op == "in" and isinstance(filter_clause.value, list):
            score += len(filter_clause.value) * self.weights["in_filter"]

        # BETWEEN is slightly more expensive
        if filter_clause.op == "between":
            score += self.weights["between_filter"]

        return score

    def _score_include(self, include: IncludeClause) -> int:
        """Score a single include clause."""
        score = self.weights["per_include"]

        # Fields in the include
        if include.select:
            score += len(include.select) * self.weights["per_field"]

        # Filters in the include
        if include.where:
            for f in include.where:
                score += self._score_filter(f)

        return score


class BudgetEnforcer:
    """
    Enforces budget limits on queries.

    Checks various limits and raises appropriate errors when exceeded.
    """

    def __init__(self, budget: Budget, scorer: ComplexityScorer | None = None) -> None:
        self.budget = budget
        self.scorer = scorer or ComplexityScorer()

    def enforce(self, request: QueryRequest) -> None:
        """
        Enforce all budget limits on a query request.

        Raises QueryBudgetExceededError if any limit is exceeded.
        """
        self._check_row_limit(request)
        self._check_field_limit(request)
        self._check_include_limit(request)
        self._check_complexity(request)

    def _check_row_limit(self, request: QueryRequest) -> None:
        """Check if requested rows exceed limit."""
        if request.take > self.budget.max_rows:
            raise QueryBudgetExceededError(
                budget_type="max_rows",
                limit=self.budget.max_rows,
                requested=request.take,
            )

    def _check_field_limit(self, request: QueryRequest) -> None:
        """Check if selected fields exceed limit."""
        if request.select and len(request.select) > self.budget.max_select_fields:
            raise QueryBudgetExceededError(
                budget_type="max_select_fields",
                limit=self.budget.max_select_fields,
                requested=len(request.select),
            )

    def _check_include_limit(self, request: QueryRequest) -> None:
        """Check if includes exceed depth limit."""
        if request.include and len(request.include) > self.budget.max_includes_depth:
            raise QueryBudgetExceededError(
                budget_type="max_includes_depth",
                limit=self.budget.max_includes_depth,
                requested=len(request.include),
            )

    def _check_complexity(self, request: QueryRequest) -> None:
        """Check if query complexity exceeds limit."""
        score = self.scorer.score(request)
        if score > self.budget.max_complexity_score:
            raise QueryBudgetExceededError(
                budget_type="complexity_score",
                limit=self.budget.max_complexity_score,
                requested=score,
            )

    def get_effective_limit(self, requested: int | None) -> int:
        """
        Get the effective row limit considering budget.

        Returns the minimum of requested limit and budget limit.
        """
        if requested is None:
            return self.budget.max_rows
        return min(requested, self.budget.max_rows)

File: packages/ormai-ts/src/policy/budgets.ts
/**
 * Budget enforcement and complexity scoring.
 *
 * Budgets prevent runaway queries by limiting rows, includes, fields,
 * and query complexity.
 */

import type { FilterClause, IncludeClause, QueryRequest } from '../core/dsl.js';
import { QueryBudgetExceededError } from '../core/errors.js';
import type { Budget } from './models.js';

/**
 * Default weights for complexity factors.
 */
export const DEFAULT_COMPLEXITY_WEIGHTS = {
  base: 1,
  perField: 1,
  perFilter: 2,
  perInclude: 10,
  perOrder: 1,
  stringFilter: 3, // contains, startswith, endswith
  inFilter: 2, // per item in IN clause
  betweenFilter: 2,
} as const;

export type ComplexityWeights = typeof DEFAULT_COMPLEXITY_WEIGHTS;

/**
 * Scores query complexity based on various factors.
 *
 * Higher scores indicate more complex (and potentially expensive) queries.
 */
export class ComplexityScorer {
  private readonly weights: ComplexityWeights;

  constructor(weights?: Partial<ComplexityWeights>) {
    this.weights = { ...DEFAULT_COMPLEXITY_WEIGHTS, ...weights };
  }

  /**
   * Calculate complexity score for a query request.
   *
   * Returns an integer score where higher = more complex.
   */
  score(request: QueryRequest): number {
    let score = this.weights.base;

    // Field selection
    if (request.select) {
      score += request.select.length * this.weights.perField;
    }

    // Filters
    if (request.where) {
      for (const f of request.where) {
        score += this.scoreFilter(f);
      }
    }

    // Includes
    if (request.include) {
      for (const inc of request.include) {
        score += this.scoreInclude(inc);
      }
    }

    // Ordering
    if (request.orderBy) {
      score += request.orderBy.length * this.weights.perOrder;
    }

    return score;
  }

  /**
   * Score a single filter clause.
   */
  private scoreFilter(filterClause: FilterClause): number {
    let score = this.weights.perFilter;

    // String operations are more expensive
    if (['contains', 'startswith', 'endswith'].includes(filterClause.op)) {
      score += this.weights.stringFilter;
    }

    // IN clauses scale with list size
    if (filterClause.op === 'in' && Array.isArray(filterClause.value)) {
      score += filterClause.value.length * this.weights.inFilter;
    }

    // BETWEEN is slightly more expensive
    if (filterClause.op === 'between') {
      score += this.weights.betweenFilter;
    }

    return score;
  }

  /**
   * Score a single include clause.
   */
  private scoreInclude(include: IncludeClause): number {
    let score = this.weights.perInclude;

    // Fields in the include
    if (include.select) {
      score += include.select.length * this.weights.perField;
    }

    // Filters in the include
    if (include.where) {
      for (const f of include.where) {
        score += this.scoreFilter(f);
      }
    }

    return score;
  }
}

/**
 * Enforces budget limits on queries.
 *
 * Checks various limits and raises appropriate errors when exceeded.
 */
export class BudgetEnforcer {
  private readonly scorer: ComplexityScorer;

  constructor(
    private readonly budget: Budget,
    scorer?: ComplexityScorer
  ) {
    this.scorer = scorer ?? new ComplexityScorer();
  }

  /**
   * Enforce all budget limits on a query request.
   *
   * Throws QueryBudgetExceededError if any limit is exceeded.
   */
  enforce(request: QueryRequest): void {
    this.checkRowLimit(request);
    this.checkFieldLimit(request);
    this.checkIncludeLimit(request);
    this.checkComplexity(request);
  }

  /**
   * Check if requested rows exceed limit.
   */
  private checkRowLimit(request: QueryRequest): void {
    if (request.take > this.budget.maxRows) {
      throw new QueryBudgetExceededError('rows', this.budget.maxRows, request.take);
    }
  }

  /**
   * Check if selected fields exceed limit.
   */
  private checkFieldLimit(request: QueryRequest): void {
    if (request.select && request.select.length > this.budget.maxSelectFields) {
      throw new QueryBudgetExceededError(
        'fields',
        this.budget.maxSelectFields,
        request.select.length
      );
    }
  }

  /**
   * Check if includes exceed depth limit.
   */
  private checkIncludeLimit(request: QueryRequest): void {
    if (request.include && request.include.length > this.budget.maxIncludesDepth) {
      throw new QueryBudgetExceededError(
        'includes',
        this.budget.maxIncludesDepth,
        request.include.length
      );
    }
  }

  /**
   * Check if query complexity exceeds limit.
   */
  private checkComplexity(request: QueryRequest): void {
    const score = this.scorer.score(request);
    if (score > this.budget.maxComplexityScore) {
      throw new QueryBudgetExceededError('complexity', this.budget.maxComplexityScore, score);
    }
  }

  /**
   * Get the effective row limit considering budget.
   *
   * Returns the minimum of requested limit and budget limit.
   */
  getEffectiveLimit(requested?: number): number {
    if (requested === undefined) {
      return this.budget.maxRows;
    }
    return Math.min(requested, this.budget.maxRows);
  }
}

/**
 * Create a complexity scorer with optional custom weights.
 */
export function createComplexityScorer(weights?: Partial<ComplexityWeights>): ComplexityScorer {
  return new ComplexityScorer(weights);
}

/**
 * Create a budget enforcer for a budget.
 */
export function createBudgetEnforcer(budget: Budget, scorer?: ComplexityScorer): BudgetEnforcer {
  return new BudgetEnforcer(budget, scorer);
}
